libdrgn/program.c

```diff
@@ -678,6 +678,30 @@ static struct drgn_error *get_prstatus_pid(struct drgn_program *prog, const char
 	return NULL;
 }
 
+static struct drgn_error *get_prpsinfo_pid(struct drgn_program *prog, const char *data,
+					   size_t size, uint32_t *ret)
+{
+	bool is_64_bit, bswap;
+	struct drgn_error *err = drgn_program_is_64_bit(prog, &is_64_bit);
+	if (err)
+		return err;
+	err = drgn_program_bswap(prog, &bswap);
+	if (err)
+		return err;
+
+	size_t offset = is_64_bit ? 24 : 20;
```

Comment:
24 looks correct for 64-bit, but on 32-bit, `unsigned long` is 4 bytes and `__kernel_uid_t` and `__kernel_gid_t` are `unsigned short` (2 bytes), so I believe the offsets are

```
struct elf_prpsinfo
{
	char	pr_state;	/* 0 */
	char	pr_sname;	/* 1 */
	char	pr_zomb;	/* 2 */
	char	pr_nice;	/* 3 */
	unsigned long pr_flag;	/* 4 */
	__kernel_uid_t	pr_uid;	/* 8 */
	__kernel_gid_t	pr_gid; /* 10 */
	pid_t	pr_pid /* 12 */, pr_ppid, pr_pgrp, pr_sid;
	/* Lots missing */
	/*
	 * The hard-coded 16 is derived from TASK_COMM_LEN, but it can't be
	 * changed as it is exposed to userspace. We'd better make it hard-coded
	 * here.
	 */
	char	pr_fname[16];	/* filename of executable */
	char	pr_psargs[ELF_PRARGSZ];	/* initial part of arg list */
};
```

Running on my i386 VM corroborates that:

```
root@vm-i386:~/drgn# python3 -Bm drgn
drgn 0.0.14+unknown (using Python 3.9.2, elfutils 0.183, without libkdumpfile)
For help, type help(drgn).
>>> import drgn
>>> from drgn import NULL, Object, cast, container_of, execscript, offsetof, reinterpret, sizeof
>>> from drgn.helpers.linux import *
>>> offsetof(prog.type("struct elf_prpsinfo"), "pr_pid")
12
```

Reply:
Fixed, need to setup 32bit VM to do proper testing myself